File: revision/8_may_24/tree_ds/bst.cpp
#include <cstddef>
#include <iostream>
#include <queue>
#include <vector>
#include <cassert>

using namespace std;

class BSTNode{
    public:
        int val;
        BSTNode* left;
        BSTNode* right;
        BSTNode(int v): val{v}, left{NULL}, right{NULL} {}
};

class BST{
    private:
        BSTNode* root;

        BSTNode* insert(BSTNode* node, int value){
            if (node == NULL) return new BSTNode(value);

            if (value < node->val) node->left = insert(node->left, value);
            else if (value > node->val) node->right = insert(node->right, value);
            return node;
        }

        int smallest_value(BSTNode* node){
            if (node->left == NULL) return node->val;
            return smallest_value(node->left);
        }

        BSTNode* remove(BSTNode* node, int value){
            if (node == NULL) return NULL;
            if (node->val == value){
                if (node->left == NULL && node->right == NULL){
                    delete node;
                    return NULL;
                }else if (node->left == NULL){
                    BSTNode* next_node = node->right;
                    delete node;
                    return next_node;
                }else if (node->right == NULL){
                    BSTNode* next_node = node->left;
                    delete node;
                    return next_node;
                }else {
                    int new_value = smallest_value(node->right);
                    node->val = new_value;
                    node->right = remove(node->right, new_value);
                }
            }else if (value < node->val) node->left = remove(node->left, value);
            else node->right = remove(node->right, value);
            return node;
        }

        void pre_order(BSTNode* node, vector<int>& out) const {
            if (node == NULL) return;
            out.push_back(node->val);
            pre_order(node->left, out);
            pre_order(node->right, out);
        }

        void in_order(BSTNode* node, vector<int>& out) const {
            if (node == NULL) return;
            in_order(node->left, out);
            out.push_back(node->val);
            in_order(node->right, out);
        }

        void post_order(BSTNode* node, vector<int>& out) const {
            if (node == NULL) return;
            post_order(node->left, out);
            post_order(node->right, out);
            out.push_back(node->val);
        }

        bool exists(BSTNode* node, int value) const {
            if (node == NULL) return false;
            if (node->val == value) return true;
            else if (value < node->val) return exists(node->left, value);
            return exists(node->right, value);
        }
    public:
        BST() : root{NULL} {}

        void insert(int val){
            root = insert(root, val);
        }

        void remove(int val){
            root = remove(root, val);
        }

        vector<int> pre_order() const {
            vector<int> out;
            pre_order(root, out);
            return out;
        }
        vector<int> in_order() const {
            vector<int> out;
            in_order(root, out);
            return out;
        }
        vector<int> post_order() const {
            vector<int> out;
            post_order(root, out);
            return out;
        }
        vector<int> bfs() const {
            vector<int> out;
            queue<BSTNode*> q;
            q.push(root);
            out.push_back(root->val);
            while (!q.empty()){
                auto node = q.front();q.pop();
                if (node->left != NULL){
                    out.push_back(node->left->val);
                    q.push(node->left);
                }
                if (node->right != NULL){
                    out.push_back(node->right->val);
                    q.push(node->right);
                }
            }
            return out;
        }
        bool exists(int val) const {
            return exists(root, val);
        }
};

int main() {
    // Create a binary search tree
    BST bst;

    // Test insert() function
    bst.insert(10);
    bst.insert(5);
    bst.insert(15);
    bst.insert(3);
    bst.insert(7);
    bst.insert(12);
    bst.insert(17);

    // Test pre_order() function
    vector<int> preOrderResult = bst.pre_order();
    assert(preOrderResult == vector<int>({10, 5, 3, 7, 15, 12, 17}));

    // Test in_order() function
    vector<int> inOrderResult = bst.in_order();
    assert(inOrderResult == vector<int>({3, 5, 7, 10, 12, 15, 17}));

    // Test post_order() function
    vector<int> postOrderResult = bst.post_order();
    assert(postOrderResult == vector<int>({3, 7, 5, 12, 17, 15, 10}));

    // Test bfs() function
    vector<int> bfsResult = bst.bfs();
    assert(bfsResult == vector<int>({10, 5, 15, 3, 7, 12, 17}));

    // Test exists() function
    assert(bst.exists(5));
    assert(!bst.exists(8));

    bst.remove(5); // Remove node with value 5
    assert(!bst.exists(5)); // Check if 5 is removed
    assert(bst.pre_order() == vector<int>({10, 7, 3, 15, 12, 17}));

    return 0;
}

File: revision/29_jun_24/src/tree/bst.rs
use std::collections::VecDeque;

struct TreeNode {
    val: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

struct Tree {
    root: Option<Box<TreeNode>>,
}

impl Tree {
    fn new() -> Self {
        Self { root: None }
    }

    fn is_empty(&self) -> bool {
        self.root.is_none()
    }

    fn has(&self, val: i32) -> bool {
        Tree::_has(self.root.as_ref(), val)
    }

    fn _has(node: Option<&Box<TreeNode>>, val: i32) -> bool {
        match node {
            None => false,
            Some(node) => {
                if node.val == val {
                    true
                } else if val < node.val {
                    Tree::_has(node.left.as_ref(), val)
                } else {
                    Tree::_has(node.right.as_ref(), val)
                }
            }
        }
    }

    fn insert(&mut self, val: i32) {
        self.root = Tree::_insert(self.root.take(), val);
    }

    fn _insert(node: Option<Box<TreeNode>>, val: i32) -> Option<Box<TreeNode>> {
        match node {
            None => Some(Box::new(TreeNode {
                val,
                left: None,
                right: None,
            })),
            Some(mut node) => {
                if val < node.val {
                    node.left = Tree::_insert(node.left.take(), val);
                } else if val > node.val {
                    node.right = Tree::_insert(node.right.take(), val)
                }
                Some(node)
            }
        }
    }

    fn remove(&mut self, val: i32) {
        self.root = Tree::_remove(self.root.take(), val);
    }

    fn _remove(node: Option<Box<TreeNode>>, val: i32) -> Option<Box<TreeNode>> {
        match node {
            None => node,
            Some(mut node) => {
                if val == node.val {
                    if node.left.is_none() && node.right.is_none() {
                        return None;
                    } else if node.left.is_none() {
                        return node.right;
                    } else if node.right.is_none() {
                        return node.left;
                    } else {
                        let left_greatest = Tree::_find_greatest(node.left.as_ref().unwrap());
                        node.val = left_greatest;
                        node.left = Tree::_remove(node.left.take(), left_greatest);
                    }
                } else if val < node.val {
                    node.left = Tree::_remove(node.left.take(), val);
                } else {
                    node.right = Tree::_remove(node.right.take(), val);
                }
                Some(node)
            }
        }
    }

    fn _find_greatest(node: &Box<TreeNode>) -> i32 {
        match node.right.as_ref() {
            None => node.val,
            Some(right) => Tree::_find_greatest(right),
        }
    }

    // give sorted array
    fn in_order(&self) -> Vec<i32> {
        let mut out = Vec::new();
        Tree::_in_order(self.root.as_ref(), &mut out);
        out
    }

    fn _in_order(node: Option<&Box<TreeNode>>, out: &mut Vec<i32>) {
        if let Some(node) = node {
            Tree::_in_order(node.left.as_ref(), out);
            out.push(node.val);
            Tree::_in_order(node.right.as_ref(), out);
        }
    }

    fn pre_order(&self) -> Vec<i32> {
        let mut out = Vec::new();
        Tree::_pre_order(self.root.as_ref(), &mut out);
        out
    }

    fn _pre_order(node: Option<&Box<TreeNode>>, out: &mut Vec<i32>) {
        if let Some(node) = node {
            out.push(node.val);
            Tree::_pre_order(node.left.as_ref(), out);
            Tree::_pre_order(node.right.as_ref(), out);
        }
    }

    fn post_order(&self) -> Vec<i32> {
        let mut out = Vec::new();
        Tree::_post_order(self.root.as_ref(), &mut out);
        out
    }

    fn _post_order(node: Option<&Box<TreeNode>>, out: &mut Vec<i32>) {
        if let Some(node) = node {
            Tree::_post_order(node.left.as_ref(), out);
            Tree::_post_order(node.right.as_ref(), out);
            out.push(node.val);
        }
    }

    fn bfs(&self) -> Vec<i32> {
        let mut queue: VecDeque<&Box<TreeNode>> = VecDeque::new();
        if let Some(node) = self.root.as_ref() {
            queue.push_back(node);
        }
        let mut out = Vec::new();
        while let Some(node) = queue.pop_front() {
            out.push(node.val);
            if let Some(left) = node.left.as_ref() {
                queue.push_back(left);
            }
            if let Some(right) = node.right.as_ref() {
                queue.push_back(right);
            }
        }
        out
    }
}

#[cfg(test)]
mod tests {
    use super::Tree;

    #[test]
    fn tree_insert() {
        let mut tree = Tree::new();

        assert!(tree.is_empty());
        assert!(!tree.has(-1));

        tree.insert(5);
        tree.insert(0);
        tree.insert(10);
        assert!(tree.has(10));
        assert!(tree.has(5));
        assert!(tree.has(0));
        assert!(!tree.has(9));
    }

    #[test]
    fn tree_remove() {
        let mut tree = Tree::new();

        let node_v = [5, 0, 10, -1, 2, 7, 11];
        for v in node_v {
            tree.insert(v);
        }

        // remove leaf
        tree.remove(7);
        assert!(!tree.has(7));

        // remove node with single child
        tree.remove(10);
        assert!(!tree.has(10));

        // remove node with two child
        tree.remove(0);
        assert!(!tree.has(0));

        assert_eq!(tree.in_order(), vec![-1, 2, 5, 11]);
    }

    #[test]
    fn tree_traversal() {
        let mut tree = Tree::new();

        let node_v = [5, 0, 10, -1, 2, 7, 11];
        for v in node_v {
            tree.insert(v);
        }

        assert_eq!(tree.in_order(), vec![-1, 0, 2, 5, 7, 10, 11]);
        assert_eq!(tree.pre_order(), vec![5, 0, -1, 2, 10, 7, 11]);
        assert_eq!(tree.post_order(), vec![-1, 2, 0, 7, 11, 10, 5]);
        assert_eq!(tree.bfs(), vec![5, 0, 10, -1, 2, 7, 11]);
    }
}
